realm/realm-library/src/main/java/io/realm/internal/Context.java

```diff
@@ -16,32 +16,60 @@
 
 package io.realm.internal;
 
-import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 
 public class Context {
 
+    // Pool to hold the phantom references.
+    // The size of array for storing phantom references will never decrease. Instead, we use another array to hold the
+    // index of the free slot. When adding the reference, pick the last index from freeIndexList and put the reference
+    // to the corresponding slot.
+    private static class ReferencesPool {
+        ArrayList<NativeObjectReference> arrayList = new ArrayList<>();
+        ArrayList<Integer> freeIndexList = new ArrayList<Integer>();
```

Comment:
One more thing, i think @zaki50 or someone mentioned `ArrayDeque` somewhere, that could optimize this even more. I will try to benchmark that.